### PR TITLE
Refactor no-index/canonical logic; Close #279

### LINKED ISSUE #279: Refactor noindex/canonical logic
https://github.com/cncf/landscapeapp/blob/master/src/utils/syncToUrl.js seems to work correctly but is really hard to follow. Among other things, you're defining existingCanonical and existingMeta twice.

Please refactor most code into functions like isCanonical, addCanonical, removeCanonical, addNoIndex, and removeNoIndex.

IsCanonical should equal yes if URL is one of the landscapes OR is selected on one of the landscapes OR is vanilla card mode. The other four functions should work idempotently. So, addCanonical should only add if it's not already there, and removeCanonical should only remove if it exists.

Then, I think the basic logic is:

If IsCanonical:
  addCanonical
  removeNoIndex
Else:
  addNoIndex
  removeCanonical
End

### PR DESCRIPTION
**Pull request by jordinl83**

### CODE REVIEW

**Comment by jordinl83** [comment]
Refactored to use [React Portals](https://reactjs.org/docs/portals.html) instead.

LGTM

**Comment by dankohn** [comment]
At the top of `head`, every page should either have meta `canonical` or `noindex`. Your implementation doesn't do that. I'm using Inspector to look at https://deploy-preview-370--landscapeapp.netlify.com/cncf/

**Comment by jordinl83** [comment]
@dankohn it's appended at the bottom, that's actually how Portals work.

 I researched about `<link rel="canonical">` and `<meta name="robots">` and I could not find anywhere that said they should be placed at the top of the `<head>`, as far as I can see they just need to be in the `<head>`.

**Comment by dankohn** [comment]
Is it possible to keep the message at the top like we have it now? It makes debugging simpler.

**Comment by jordinl83** [comment]
@dankohn it's definitely possible, but we would have to manually update the DOM instead of allowing React do it. So the `Head` class would look like this:

```js
class Head extends Component {
  canonicalTag = () => {
    let el = document.createElement("link");
    el.rel = "canonical";
    el.href = window.location.href;
    return el;
  }

  noIndexTag = () => {
    let el = document.createElement("meta");
    el.name = "robots";
    el.content = "noindex";
    return el;
  }

  componentWillReceiveProps (nextProps, nextContext) {
    const { location } = this.props;
    const { head } = document;

    if (location.pathname != nextProps.location.pathname) {
      ["link[rel='canonical']", "meta[name='robots']"].forEach((selector) => {
        const element = head.querySelector(selector);
        element && element.remove();
      })

      const indexTag = isCanonical(nextProps.location.pathname) ? this.canonicalTag() : this.noIndexTag();
      head.insertBefore(indexTag, head.firstChild);
    }
  }

  render() {
    return null;
  }
};
```

Instead of:

```js
const Head = ({ location }) => {
  const indexTag = isCanonical(location.pathname) ?
    <link rel="canonical" href={window.location.href}/> :
    <meta name="robots" content="noindex"/>;

  return ReactDOM.createPortal(indexTag, document.head);
};
```

Not sure the additional complexity is worth it. Thoughts?

**Comment by jordinl83** [comment]
@dankohn found this extension that shows `head` contents: https://chrome.google.com/webstore/detail/meta-seo-inspector/ibkclpciafdglkjkcibmohobjkcfkaef?hl=en
